File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/camera2/TotalCaptureResultImpl.kt
package dev.zeekr.camerax_android.camera2

import android.hardware.camera2.CaptureResult
import android.hardware.camera2.TotalCaptureResult
import android.os.Build
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiTotalCaptureResultProxyApi

class TotalCaptureResultImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiTotalCaptureResultProxyApi(registrar) {
    override fun getPartialResults(pigeon_instance: TotalCaptureResult): List<CaptureResult> {
        return pigeon_instance.partialResults
    }

    override fun getPhysicalCameraResults(pigeon_instance: TotalCaptureResult): Map<String, CaptureResult> {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            return pigeon_instance.physicalCameraResults
        } else {
            throw UnsupportedOperationException("Call requires API level 28")
        }
    }

    override fun getPhysicalCameraTotalResults(pigeon_instance: TotalCaptureResult): Map<String, TotalCaptureResult> {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return pigeon_instance.physicalCameraTotalResults
        } else {
            throw UnsupportedOperationException("Call requires API level 31")
        }
    }
}

File: camerax_android/lib/src/camera2/total_capture_result_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/camera2.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class TotalCaptureResultImpl extends CaptureResultImpl
    implements TotalCaptureResult {
  @override
  TotalCaptureResultProxyApi get api {
    final api = super.api;
    if (api is! TotalCaptureResultProxyApi) throw TypeError();
    return api;
  }

  TotalCaptureResultImpl.internal(TotalCaptureResultProxyApi super.api)
    : super.internal();

  @override
  Future<List<CaptureResult>> getPartialResults() =>
      api.getPartialResults().then((e) => e.map((e) => e.impl).toList());

  @override
  Future<Map<String, CaptureResult>> getPhysicalCameraResults() => api
      .getPhysicalCameraResults()
      .then((e) => e.map((k, v) => MapEntry(k, v.impl)));

  @override
  Future<Map<String, TotalCaptureResult>> getPhysicalCameraTotalResults() => api
      .getPhysicalCameraTotalResults()
      .then((e) => e.map((k, v) => MapEntry(k, v.impl)));
}

extension TotalCaptureResultProxyApiX on TotalCaptureResultProxyApi {
  TotalCaptureResult get impl => TotalCaptureResultImpl.internal(this);
}
